api/diff.go

```diff
@@ -30,77 +30,72 @@ func apiDiffHandler(w http.ResponseWriter, r *http.Request) {
+		return nil, err
+	} else if len(runIDs) > 0 {
+		runs, err := runIDs.LoadTestRuns(store)
+		// If all errors are NoSuchEntity, we don't treat it as an error.
+		// If err is nil, the type conversion will fail.
+		if multiError, ok := err.(appengine.MultiError); ok {
+			all404s := true
+			for _, err := range multiError {
+				if err != datastore.ErrNoSuchEntity {
+					all404s = false
+					break
+				}
+			}
+			if all404s {
+				return nil, nil
+			}
+		}
+		// err many be non nil here.
+		return runs, err
+	}
+
+	// NOTE: We use the same params as /results, but also support
+	// 'before' and 'after' and 'filter'.
+	runFilter, err := shared.ParseTestRunFilterParams(q)
+	if err != nil {
+		return nil, err
+	}
+	if beforeAndAfter, err := shared.ParseBeforeAndAfterParams(q); err != nil {
+		return nil, err
+	} else if len(beforeAndAfter) > 0 {
+		runFilter.Products = beforeAndAfter
+	}
+	runsByProduct, err := LoadTestRunsForFilters(store, runFilter)
+	if err != nil {
+		return nil, err
+	}
+	return runsByProduct.AllRuns(), nil
 }
 
 func handleAPIDiffGet(w http.ResponseWriter, r *http.Request) {
 	ctx := shared.NewAppEngineContext(r)
-
+	store := shared.NewAppEngineDatastore(ctx, true)
 	q := r.URL.Query()
-	runIDs, err := shared.ParseRunIDsParam(q)
+
+	runs, err := loadDiffRuns(store, q)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
-
-	var diffFilter shared.DiffFilterParam
-	var paths mapset.Set
-	if diffFilter, paths, err = shared.ParseDiffFilterParams(q); err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
+	if runs == nil {
+		http.NotFound(w, r)
 		return
 	}
 
-	store := shared.NewAppEngineDatastore(ctx, true)
-	var runs shared.TestRuns
-	if len(runIDs) > 0 {
-		runs, err = runIDs.LoadTestRuns(store)
-		if err != nil {
-			if multiError, ok := err.(appengine.MultiError); ok {
-				all404s := true
-				for _, err := range multiError {
-					if err != datastore.ErrNoSuchEntity {
-						all404s = false
-					}
-				}
-				if all404s {
-					http.NotFound(w, r)
-					return
-				}
-			}
-		}
-	} else {
-		// NOTE: We use the same params as /results, but also support
-		// 'before' and 'after' and 'filter'.
-		runFilter, parseErr := shared.ParseTestRunFilterParams(q)
-		if parseErr != nil {
-			http.Error(w, err.Error(), http.StatusBadRequest)
-			return
-		}
-		beforeAndAfter, parseErr := shared.ParseBeforeAndAfterParams(q)
-		if parseErr != nil {
-			http.Error(w, err.Error(), http.StatusBadRequest)
-			return
-		}
-		if len(beforeAndAfter) > 0 {
-			runFilter.Products = beforeAndAfter
-		}
-		var runsByProduct shared.TestRunsByProduct
-		runsByProduct, err = LoadTestRunsForFilters(store, runFilter)
-		if err != nil {
-			runs = runsByProduct.AllRuns()
-		}
+	if len(runs) != 2 {
+		http.Error(w, fmt.Sprintf("Diffing requires exactly 2 runs, but found %v", len(runs)), http.StatusBadRequest)
+		return
 	}
 
+	diffFilter, paths, err := shared.ParseDiffFilterParams(q)
```

Comment:
I think I would rather have it in the beginning of the handler, so if this checks fails, it doesn't do all the extra work necessarily (e.g. `LoadTestRuns`). And it might be more clearer on the param requirements

Reply:
Moved this along with the error handling to the beginning.